File: app/src/main/java/farasense/mobile/service/download/DownloadFaraSenseSensorService.java
package farasense.mobile.service.download;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.os.Message;
import android.os.Process;
import android.util.Log;

import java.util.List;
import java.util.Date;
import farasense.mobile.api.RestClient;
import farasense.mobile.api.base.ErrorListener;
import farasense.mobile.api.base.RestError;
import farasense.mobile.api.base.SuccessListener;
import farasense.mobile.model.DAO.FaraSenseSensorDAO;
import farasense.mobile.model.DAO.base.BaseDAO;
import farasense.mobile.model.realm.FaraSenseSensor;
import farasense.mobile.service.base.BaseService;
import farasense.mobile.service.listener.OnDownloadContentListener;
import farasense.mobile.util.DateUtil;
import farasense.mobile.util.ConnectionUtil;
import farasense.mobile.util.PermissionUtil;

public class DownloadFaraSenseSensorService extends BaseService {

    private final class FaraSenseSensorServiceHandler extends Handler {

        public FaraSenseSensorServiceHandler(Looper looper) {
            super(looper);
        }

        @Override
        public void handleMessage(final Message msg) {
            while (!shouldStop) {
                try {
                    if (ConnectionUtil.isDataConnectionAvailable(getApplicationContext())) {
                        RestClient.getInstance().getFaraSenseSensor(new SuccessListener<List<FaraSenseSensor>>() {
                            @Override
                            public void onSuccess(List<FaraSenseSensor> response) {
                                Log.d("FARASENSE SENSOR", LOG_DSERVICE_OK);
                                FaraSenseSensorDAO.saveFromServer(response);
                                if (PermissionUtil.hasPermissions(getApplicationContext(), PermissionUtil.PERMISSIONS)) {
                                    BaseDAO.saveDataBase(getApplicationContext());
                                }
                            }
                        }, new ErrorListener() {
                            @Override
                            public void onError(RestError restError) {
                                Log.d("ERROR FARASENSE SENSOR", LOG_DSERVICE_ERROR);
                            }
                        }, DateUtil.getFirstMomentOfTheDay()
                        , DateUtil.getNow());
                        Thread.sleep(SLEEP_WAIT);
                    } else {
                        Log.d("ERROR FARASENSE SENSOR", LOG_DSERVICE_ERROR);
                        Thread.sleep(SLEEP_TRY_AGAIN);
                    }
                }
                catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
            stopSelf(msg.arg1);
        }
    }

    public static void download(OnDownloadContentListener onDownloadContentListener, Date startDate, Date finalDate) {
        new Thread(() -> RestClient.getInstance().getFaraSenseSensor(new SuccessListener<List<FaraSenseSensor>>() {
            @Override
            public void onSuccess(List<FaraSenseSensor> response) {
                Log.d("FARASENSE SENSOR", LOG_DSERVICE_OK);
                onDownloadContentListener.onSucess();
                FaraSenseSensorDAO.saveFromServer(response);
            }
        }, new ErrorListener() {
            @Override
            public void onError(RestError restError) {
                Log.d("ERROR FARASENSE SENSOR", LOG_DSERVICE_ERROR);
                onDownloadContentListener.onFail();
            }
        }, startDate, finalDate)).start();
    }

    @Override
    public void onCreate() {
        HandlerThread threadFaraSenseSensor = new HandlerThread("FaraSenseSensorStartArguments", Process.THREAD_PRIORITY_BACKGROUND);
        threadFaraSenseSensor.start();
        looper = threadFaraSenseSensor.getLooper();
        handler = new FaraSenseSensorServiceHandler(looper);
    }
}

File: app/src/main/java/farasense/mobile/service/download/DownloadFaraSenseSensorService.kt
package farasense.mobile.service.download

import android.content.Context
import android.os.*
import android.util.Log
import farasense.mobile.BuildConfig
import farasense.mobile.api.RestClient
import farasense.mobile.api.base.ErrorListener
import farasense.mobile.api.base.RestError
import farasense.mobile.api.base.SuccessListener
import farasense.mobile.model.DAO.FaraSenseSensorDAO
import farasense.mobile.model.DAO.base.BaseDAO
import farasense.mobile.model.realm.FaraSenseSensor
import farasense.mobile.service.base.BaseService
import farasense.mobile.service.listener.OnDownloadContentListener
import farasense.mobile.util.ConnectionUtil
import farasense.mobile.util.DateUtil
import farasense.mobile.util.PermissionUtil
import farasense.mobile.view_model.base.BaseObservableViewModel
import java.util.*

class DownloadFaraSenseSensorService : BaseService() {

    private inner class FaraSenseSensorServiceHandler(looper: Looper) : Handler(looper) {

        override fun handleMessage(msg: Message) {
            while (!shouldStop) {
                try {
                    if (BaseObservableViewModel.serviceStarted && ConnectionUtil.isDataConnectionAvailable(applicationContext)) {
                        RestClient.getFaraSenseSensor(applicationContext, object : SuccessListener<List<FaraSenseSensor>>() {
                            override fun onSuccess(response: List<FaraSenseSensor>) {
                                Log.d("FARASENSE SENSOR", LOG_DSERVICE_OK)
                                FaraSenseSensorDAO().saveFromServer(response)
                                if (BuildConfig.DEBUG && PermissionUtil.hasPermissions(applicationContext, *PermissionUtil.PERMISSIONS_DEBUG)) {
                                    BaseDAO().saveDataBase(applicationContext)
                                }
                            }
                        }, object : ErrorListener() {
                            override fun onError(restError: RestError) {
                                Log.d("ERROR FARASENSE SENSOR", LOG_DSERVICE_ERROR)
                            }
                        }, DateUtil.firstMomentOfTheDay, DateUtil.now)
                        Thread.sleep(BaseService.SLEEP_WAIT.toLong())
                    } else {
                        Log.d("ERROR FARASENSE SENSOR", LOG_DSERVICE_ERROR)
                        Thread.sleep(BaseService.SLEEP_TRY_AGAIN.toLong())
                    }
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                }

            }
            stopSelf(msg.arg1)
        }
    }

    override fun onCreate() {
        val threadFaraSenseSensor = HandlerThread("FaraSenseSensorStartArguments", Process.THREAD_PRIORITY_BACKGROUND)
        threadFaraSenseSensor.start()
        looper = threadFaraSenseSensor.looper
        handler = FaraSenseSensorServiceHandler(looper!!)
    }

    companion object {

        fun download(context: Context, onDownloadContentListener: OnDownloadContentListener, startDate: Date, finalDate: Date) {
            Thread {
                RestClient.getFaraSenseSensor(context, object : SuccessListener<List<FaraSenseSensor>>() {
                    override fun onSuccess(response: List<FaraSenseSensor>) {
                        Log.d("FARASENSE SENSOR", LOG_DSERVICE_OK)
                        onDownloadContentListener.onSucess()
                        FaraSenseSensorDAO().saveFromServer(response)
                    }
                }, object : ErrorListener() {
                    override fun onError(restError: RestError) {
                        Log.d("ERROR FARASENSE SENSOR", LOG_DSERVICE_ERROR)
                        onDownloadContentListener.onFail()
                    }
                }, startDate, finalDate)
            }.start()
        }
    }
}
